configmap/dynamic.go

```diff
@@ -0,0 +1,250 @@
+package configmap
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+type ConfigState struct {
+	Raw       map[string]string
+	Mapped    interface{}
+	Timestamp time.Time
+}
+
+type ConfigMapperInput struct {
+	RawCurrent map[string]string
+	Previous   ConfigState
+}
+
+// ConfigMapper parses configmap data and returns user provided object or error
+type ConfigMapper func(rawCurrent map[string]string, previous *ConfigState) (interface{}, error)
+
+type OnUpdateCallback func(current *ConfigState, previous *ConfigState)
+
+type OnErrorCallback func(err error, previous *ConfigState)
+
+type Options struct {
+	// If set to true, load data during initialization only, and do not refresh them.
+	DisableDynamicUpdates bool
+	// Update callback is called on each data evaluation both successful and failed
+	OnUpdateCallback OnUpdateCallback
+	// Called when there is an during config data processing
+	OnErrorCallback OnErrorCallback
+	// Configuration template. If set, and configmap is not found, a new one is created using this template.
+	// If configmap is there, but some parameters are missing, there are added with the default values.
+	Template map[string]string
+	// If set to true, unrecognized parameters are removed from a configmap
+	RemoveUnknownProperties bool
+}
+
+type DynamicConfig interface {
+	// GetBootstrap returns the configuration state loaded on bootstrap. As bootstrap state never changes.
+	GetBootstrap() interface{}
+
+	// GetRawBootstrap returns the configmap loaded on bootstrap. As bootstrap state never changes.
+	GetRawBootstrap() map[string]string
+
+	// Get returns the latest configuration state
+	Get() interface{}
+
+	// GetRaw returns the latest configmap
+	GetRaw() map[string]string
+
+	// Close stops the internal refresh process
+	Close()
+}
+
+type dynamicConfigInternal struct {
```

Comment:
Why do we need the two structs?
Can't we just have one struct, that has some upper case and some lower case stuff?

Reply:
`DynamicConfig` is an interface so we can easily switch implementation if needed in the future.